### ISSUE TITLE
Apparent regression in BB finding algorithm

### ISSUE THREAD

**Issue by SimonBiggs** [jrkerns/pylinac#268]
**Describe the bug**
It appears that for some cases, from pylinac version 2.2.6 --> 2.2.7 there has been a regression in the BB finding algorithm. The screenshots given below compare the [PyMedPhys](https://github.com/pymedphys/pymedphys/) wlutz algorithm with the pylinac algorithms from version 2.2.6 and 2.2.7.

**Expected behavior**
Expect a profile taken through the declared centre of the ball bearing should be symmetrical about the BB centre, therefore a flipped profile of the BB in the image should result in overlapping BB penumbras.

**Screenshots**
![PyMedPhys](https://user-images.githubusercontent.com/6559099/71410234-8ebf2f00-2698-11ea-9b10-2d10d14f4031.png)
![Pylinac v2 2 6](https://user-images.githubusercontent.com/6559099/71410235-8ff05c00-2698-11ea-8906-30a30aa431f0.png)
![Pylinac v2 2 7](https://user-images.githubusercontent.com/6559099/71410238-91218900-2698-11ea-808b-39750ab975dc.png)

**To Reproduce**

To recreate the above screenshots run the following.

```bash
pip install pymedphys==0.17.1
```

```python
import matplotlib.pyplot as plt

import pymedphys
import pymedphys._wlutz.findfield
import pymedphys._wlutz.findbb
import pymedphys._wlutz.iview
import pymedphys._wlutz.imginterp
import pymedphys._wlutz.reporting
import pymedphys._wlutz.pylinac

image_path = pymedphys.data_path("pylinac_wlutz_regression_image.jpg")

penumbra = 2
edge_lengths = [20, 24]
bb_diameter = 8

x, y, img = pymedphys._wlutz.iview.iview_image_transform(image_path)

field = pymedphys._wlutz.imginterp.create_interpolated_field(x, y, img)
initial_centre = pymedphys._wlutz.findfield._initial_centre(x, y, img)

field_centre, field_rotation = pymedphys._wlutz.findfield.field_centre_and_rotation_refining(
    field, edge_lengths, penumbra, initial_centre
)    

bb_centre = pymedphys._wlutz.findbb.optimise_bb_centre( 
    field, 
	bb_diameter, 
	edge_lengths, 
	penumbra, 
	field_centre, 
	field_rotation, 
	pylinac_tol=0.5  # required to ignore diff to pylinac in this case
    #  pylinac_tol should normally be set to about 0.1 or 0.2
)

pylinac = pymedphys._wlutz.pylinac.run_wlutz(
    field, edge_lengths, penumbra, field_centre, field_rotation)

pymedphys._wlutz.reporting.image_analysis_figure(
    x,
    y,
    img,
    bb_centre,
    field_centre,
    field_rotation,
    bb_diameter,
    edge_lengths,
    penumbra,
)

plt.title('PyMedPhys Basinhopping Method')
plt.tight_layout()
plt.savefig('PyMedPhys.png')

pymedphys._wlutz.reporting.image_analysis_figure(
    x,
    y,
    img,
    pylinac['v2.2.6']['bb_centre'],
    pylinac['v2.2.6']['field_centre'],
    field_rotation,
    bb_diameter,
    edge_lengths,
    penumbra,
)

plt.title('Pylinac v2.2.6 Filter and Profile Method')
plt.tight_layout()
plt.savefig('Pylinac v2.2.6.png')

pymedphys._wlutz.reporting.image_analysis_figure(
    x,
    y,
    img,
    pylinac['v2.2.7']['bb_centre'],
    pylinac['v2.2.7']['field_centre'],
    field_rotation,
    bb_diameter,
    edge_lengths,
    penumbra,
)

plt.title('Pylinac v2.2.7 Filter and Scikit-Image Method')
plt.tight_layout()
plt.savefig('Pylinac v2.2.7.png')
```

**Additional context**
This is using a vendored version of pylinac so that the code can run pylinac 2.2.6 and 2.2.7 at the same time while comparing the results to what PyMedPhys has determined.

Before passing the image to pylinac it is rotated such that the field has no collimator rotation.

**Comment by brjdenis** [comment]
Hi there!

Just adding something to this. I too have noticed that the BB finding algorithm may sometimes miss the geometric center of the BB for larger BBs (compared to the field), or when the contrast between the BB and the field is not large enough, or when I do the non-central WL test where the distribution of pixel values around the BB is always non-concentric.  Here is an example. On the left image with normal gray colormap it would seem that the BB center is positioned too much towards the right. But if one examines the distribution of pixel values, it turns out that the center of mass is indeed shifted outwards. On the right image I am using the "prism" colormap to show the pixel value distribution for the same image. The blue cross is the field CAX, the red cross is the BB center. I didn't do any further or better analysis, but looking at the image, I would say that the non-centrality of profiles is inherent to this algorithm. When I switched to version 2.2.7, I did some test and I didn't get results that were different by much more than 0.1 mm. I think I would get a similar result in older versions as well. But I didn't check it closely, so I can't be sure.

![wl_test_bb](https://user-images.githubusercontent.com/58114103/71446436-a68fc500-2723-11ea-86f6-1ce5fdddaca4.png)


Live long and prosper!
Denis

**Comment by SimonBiggs** [comment]
@brjdenis would you be able to upload those failing images to the PyMedPhys Zenodo file storage (https://zenodo.org/communities/pymedphys/)? I'd be keen to compare the WLutz results for those images to PyMedPhys' implementation.

**Comment by brjdenis** [comment]
@SimonBiggs sure I can. I will upload the images that are similar to the one I showed on this thread.

**Comment by brjdenis** [comment]
Hi again, I couldn't upload to zenodo. Here is instead an alternative link:
https://github.com/brjdenis/qaserver/blob/master/files/non-central-wl.zip

**Comment by SimonBiggs** [comment]
Heya @brjdenis,

I have uploaded an example notebook analysing those images to the following link:

http://www.simonbiggs.net/wlutz

If you want to run that same code locally I have also attached the notebook file to this issue:
[wlutz-dicom-example.zip](https://github.com/jrkerns/pylinac/files/4030022/wlutz-dicom-example.zip)

The PyMedPhys wlutz code is still not officially released. It needs to be put through its paces first. But if you're keen on testing it and providing me more images when it struggles I would be massively appreciative of that.

Cheers,
Simon

**Comment by brjdenis** [comment]
Hi @SimonBiggs,

I checked your results. It turns out that the difference between PyMedPhys and Pylinac calculated CAX-BB difference is less than 0.1 mm. I was honestly expecting it to be larger because to the naked eye it looks like the BB has the geometric center elsewhere. I guess I was wrong.

**Comment by brjdenis** [comment]
Hi @jrkerns @SimonBiggs it's me again.

I repeated the analysis with Pylinac and PipsPro for several old images. The result is this: mostly they agree within 0.1 mm. However, in some cases, the difference can be larger. This plot shows the average difference between the two calculations (Pylinac minus PipsPro):

![image](https://user-images.githubusercontent.com/58114103/72640424-24806e00-3968-11ea-914b-fb86a7570c3d.png)

Images acquired with 10 FFF on Elekta VersaHD and high contrast 2.5 MV images on a Varian TrueBeam show the biggest difference when compared to PipsPro. I think that Pylinac's BB searching algorithm tends to fail for these images. I uploaded the example sets so that you can test them:

[elekta_BB_problem.zip](https://github.com/jrkerns/pylinac/files/4078661/elekta_BB_problem.zip)

[varian_BB_probleml.zip](https://github.com/jrkerns/pylinac/files/4078670/varian_BB_probleml.zip)

The results are:
![image](https://user-images.githubusercontent.com/58114103/72641522-b25d5880-396a-11ea-8809-bb681a64a366.png)

For the three Elekta images, the differences are not very large, but are larger than 0.1 mm. And you can spot with the naked eye that Pylinac perhaps does not locate the center of the BB accurately. For Varian images, the results show a larger difference. For normal energies I don't see this, only for 2.5 MV.

Best regards,
Denis

**Comment by brjdenis** [comment]
Sorry, I made an error in the second table.

![image](https://user-images.githubusercontent.com/58114103/72641800-43ccca80-396b-11ea-8619-141f049eb2d6.png)

**Comment by SimonBiggs** [comment]
Just to clarify @brjdenis what version of pylinac was this?

**Comment by brjdenis** [comment]
@SimonBiggs , the latest, 2.2.7

**Comment by SimonBiggs** [comment]
Can you also repeat on version 2.2.6? Especially given this issue is to show that the previous version of `pylinac` is an improvement over the current.

**Comment by brjdenis** [comment]
Unfortunately, I was randomly selecting images, so right now it would be hard to find them. But, I will repeat the analysis, and gather the images together for comparison.

**Comment by SimonBiggs** [comment]
Yup, that'd be great. If you were keen I would also appreciate `pymedphys` to be included in that comparison if you could. It would be nice to have a representative comparison between pymedphys, pylinac 2.2.6, 2.2.7 and pipspro.

**Comment by SimonBiggs** [comment]
@brjdenis, I have just undergone an initial release of some of the Winston Lutz code. I have written up an example using the DICOM images you provided earlier:

https://docs.pymedphys.com/howto/wlutz/varian-dicom-example.html

To run that example you will need to install pymedphys with:

```bash
pip install pymedphys==0.20.0
``` 

I would be very keen if you could run your tests against PyMedPhys as well.

**Comment by brjdenis** [comment]
Hi @SimonBiggs  and @jrkerns !

I repeated the analysis with a new set of images (not the same as before). This time I collected all the data. I was not able to do the anaylsis with Pylinac 2.2.6 because of some problem with Scipy's imresize, and I also did not do the analysis with PyMedPhys because I didn't know how to set the parameters. 
But, all the data is in this zip archive, so you can redo the analysis yourselves. PipsPro results are included for each image. To run the analysis see calc.py.

https://drive.google.com/file/d/1AxpErc1IalboS6Sjo5EdRTJBJqZ1krsM/view?usp=sharing

I hope the link works.

Denis

**Comment by jrkerns** [comment]
Hi guys, 
      I'm just now digging into this. The link works Denis. I tweaked it a bit to give me a 2D plot and now It's easier to find the images that have the largest deviations visually. Will report back. 
![wl_values](https://user-images.githubusercontent.com/4803336/79032475-5d497100-7b6c-11ea-90a5-d37c32d69f41.png)

**Comment by jrkerns** [comment]
Okay, so after some digging I found two issues. First let me describe roughly what pylinac was/is trying to do. 

To find the BB, pylinac uses a double-thresholding technique to find a [boolean pattern that looks like the BB](https://github.com/jrkerns/pylinac/blob/ad10a4835580a9e8e177962254d277e44ec078dc/pylinac/winston_lutz.py#L633-L652). Then it's [weighted by the intensity](https://github.com/jrkerns/pylinac/blob/ad10a4835580a9e8e177962254d277e44ec078dc/pylinac/winston_lutz.py#L655-L658) of the image. The idea is that the center of the BB should give more weight than peripheral pixels, as the boolean mask that's determined from step one isn't perfectly round and the centroid shouldn't be influenced by distal pixels with little intensity. The idea is to find a weighted centroid using an image like this: 
![correct_intensity](https://user-images.githubusercontent.com/4803336/79035278-84ab3880-7b82-11ea-95c2-9d0cd36fb922.png)

Unfortunately, that's not what was happening. The pixel weighting was using the convention that radiation increases the pixel intensity. This would result in the *lowest* pixel intensity at the center and the highest at the edges, which were choppy: 
![incorrect_intensity](https://user-images.githubusercontent.com/4803336/79035295-b7edc780-7b82-11ea-92ee-d3c7e18c8a13.png)

Second, I finally determined that there was an issue with how the pixel values were being handled. As I recent discovered, the RescaleSlope and RescaleIntercept can 1) not exist, 2) be "dumb" values (1, 0), or 3) be correct with values that convert to CU/MU. Some of the images in Denis' set contained valid scale/intercept values while others did not. Images that do not have the tags or correct values end up having values with very, very small ranges. The edge of the image can be 15,900 and the radiation field values can be 16,200. This very small range of pixel values causes the intensity image to be essentially flat and practically boolean. This essentially means that when an image is using pixel values that the weighted intensity is not taken into account and we get an unweighted centroid. 

The following images show the agreement to Denis' dataset from PipsPro 1) before addressing this, 2) after correcting the inverted weighting intensity, and 3) after applying the correct intensity AND "stretching" images with raw pixel values to be normalized (out of field =~0 and in-field=~1). 

Before we started:
![wl_values](https://user-images.githubusercontent.com/4803336/79035411-efa93f00-7b83-11ea-98fc-596d51d153ec.png)

After intensity correction:
![wl_after_correct_inversion](https://user-images.githubusercontent.com/4803336/79035415-f932a700-7b83-11ea-97fc-98921902b258.png)

After intensity correction and weighting correction w/ raw pixels:
![wl_after_inversion_and_normalization](https://user-images.githubusercontent.com/4803336/79035418-05b6ff80-7b84-11ea-9671-4f16a4378830.png)

Pretty crazy right?

**Comment by SimonBiggs** [comment]
@jrkerns that is some awesome work right there :) Thanks James :)

**Comment by randlet** [comment]
Seconded! Very nice work James!

**Comment by brjdenis** [comment]
This is great James. Thank you.
